lib/repo/git_repository.rb

```diff
@@ -55,15 +64,72 @@ def initialize(connect_string)
     # location 'connect_string'
     def self.create(connect_string)
       if GitRepository.repository_exists?(connect_string)
-        raise RepositoryCollision.new("There is already a repository at #{connect_string}")
+        raise RepositoryCollision.new(
+                  "There is already a repository at #{connect_string}")
       end
       if File.exists?(connect_string)
         raise IOError.new("Could not create a repository at #{connect_string}:
                           some directory with same name exists already")
       end
 
-      #Create it (we're not going to use a bare repository)
-      repo = Rugged::Repository.init_at(connect_string)
+      ga_repo = Gitolite::GitoliteAdmin.new(
+        Repository.conf[:REPOSITORY_STORAGE] +
+          '/gitolite-admin', GITOLITE_SETTINGS)
+
+      # Bring the repo up to date
+      ga_repo.reload!
+
+      # Grab the gitolite admin repo config
+      conf = ga_repo.config
+
+      repo_name = File.basename(connect_string.split('/').last)
+
+      # Grab the repo in question, if it does not exist, create it
+      repo = ga_repo.config.get_repo(repo_name)
+      if repo.nil?
+        # Generate new repo since this repo hasn't been created yet
+        repo = Gitolite::Config::Repo.new(repo_name)
+      end
+
+      # Add permissions for git user
+      repo.add_permission('RW+', '', 'vagrant')
+
+      # Add the repo to the gitolite admin config
+      conf.add_repo(repo)
+
+      # Readd the 'git' public key to the gitolite admin repo after changes
+      admin_key = Gitolite::SSHKey.from_file(
+        GITOLITE_SETTINGS[:public_key])
+      ga_repo.add_key(admin_key)
+
+      # Stage and push the changes to the gitolite admin repo
+      ga_repo.save_and_apply
+
+      # Repo is created by gitolite, proceed to clone it in
+      # the repository storage location
+      cloned_repo = Git.clone(
+        'git@localhost:' + repo_name, REPOSITORY_STORAGE + '/' + repo_name)
+
+      repo = Rugged::Repository.discover(REPOSITORY_STORAGE + '/' + repo_name)
+
+      # Lets make some sample files and the new master branch
+      cloned_repo.reset
+      cloned_repo.branch('master')
+
+      # Create the README.md on the filesystem
+      file_path_for_readme = File.join(repo.workdir, 'README.md')
+      File.open(file_path_for_readme, 'w+') do |readme|
+        readme.write('Initial commit.')
+      end
+
+      cloned_repo.add(all: true)
+
+      cloned_repo.commit_all('Initial commit.')
+      cloned_repo.push
+
+      # Finished cloning repo
+
+      repo = Rugged::Repository.discover(REPOSITORY_STORAGE + '/' + repo_name)
 
       # Do an initial commit with a README to create index.
```

Comment:
Hmm okay, thanks @SoftwareDev! @reidka have been trying out the code, and it seems to more or less work. Very exciting stuff!